### PR TITLE
Remove support for Envoy 1.15

### PR DESCRIPTION
**Pull request by eculver**
Refs #11100, stacked on #11115.

This is a follow on to #11115 to remove support for Envoy 1.15. The only notable changes are around the feature flags that we had in place for Envoy < 1.16, which means _only 1.15_. As such, I am removing these flags and code that references them.

Note: As this is pretty unfamiliar territory for me, I'd like to specifically call out the bits in `./agent/xds/clusters.go` that are getting removed (`maybeInjectStubClusterForGateways` and calling code). This appears correct to me, but need someone to confirm.

I'll rebase on `main` once #11115 lands but for now this will be based on `eculver/envoy-1.19`.

### CODE REVIEW

File: agent/xds/envoy_versioning.go [hunk 1]
```diff
@@ -82,13 +73,12 @@ func determineSupportedProxyFeaturesFromVersion(version *version.Version) (suppo
...
 	sf := supportedProxyFeatures{}
 
-	if version.LessThan(minVersionAllowingEmptyGatewayClustersWithIncrementalXDS) {
-		sf.GatewaysNeedStubClusterWhenEmptyWithIncrementalXDS = true
-	}
-
-	if version.LessThan(minVersionAllowingMultipleIncrementalXDSChanges) {
-		sf.IncrementalXDSUpdatesMustBeSerial = true
-	}
+	// add version contraints to poipulate feature flags here when necessary, for example:
```
Comment (freddygv) [inline agent/xds/envoy_versioning.go:76]:
```suggestion
	// add version constraints to populate feature flags here when necessary, for example:
```

**Review by freddygv — APPROVED** [review]
LGTM, one minor suggestion

File: agent/xds/envoy_versioning_test.go [hunk 2]
```diff
@@ -109,14 +109,7 @@ func TestDetermineSupportedProxyFeaturesFromString(t *testing.T) {
 	}
```
Comment (rboyer) [inline agent/xds/envoy_versioning_test.go:109]: I've been moving older versions up here into the "errTooOld" test list,so probably best to add all of the "1.15.x" series.

File: agent/xds/clusters.go [hunk 3]
```diff
@@ -38,53 +38,24 @@ func (s *ResourceGenerator) clustersFromSnapshot(cfgSnap *proxycfg.ConfigSnapsho
...
 			return nil, err
 		}
-		return s.maybeInjectStubClusterForGateways(res)
+		return res, nil
 	default:
 		return nil, fmt.Errorf("Invalid service kind: %v", cfgSnap.Kind)
 	}
 }
 
-func (s *ResourceGenerator) maybeInjectStubClusterForGateways(resources []proto.Message) ([]proto.Message, error) {
```
Comment (rboyer) [inline agent/xds/clusters.go:59]: yay!

File: agent/xds/delta.go [hunk 4]
```diff
@@ -310,12 +310,6 @@ func (s *Server) processDelta(stream ADSDeltaStream, reqCh <-chan *envoy_discove
 				}
 				if sent {
 					sentType[op.TypeUrl] = struct{}{}
-					if generator.ProxyFeatures.IncrementalXDSUpdatesMustBeSerial {
```
Comment (rboyer) [inline agent/xds/delta.go:313]: yay!

File: agent/xds/envoy_versioning.go [hunk 5]
```diff
@@ -27,24 +27,15 @@ type unsupportedVersion struct {
...
-	// GatewaysNeedStubClusterWhenEmptyWithIncrementalXDS is needed to paper
-	// over some weird envoy behavior.
 	//
-	// For some reason Envoy versions prior to 1.16.0 when sent an empty CDS
-	// list via the incremental xDS protocol will correctly ack the message and
-	// just never request LDS resources.
-	GatewaysNeedStubClusterWhenEmptyWithIncrementalXDS bool
-
-	// IncrementalXDSUpdatesMustBeSerial is needed to avoid an envoy crash.
+	// For example, we previously had flags for Envoy < 1.16 called:
```
Comment (rboyer) [inline agent/xds/envoy_versioning.go:31]: I like this. Prior to these 2 flags existing the struct was empty for a bit. Prior to that it had completely different features available. When I went to add these two flags I had to look at my own file history to see how to re-add a feature flag.

**Comment by hc-github-team-consul-core** [comment]
:cherries: If backport labels were added before merging, cherry-picking will start automatically.

To retroactively trigger a backport after merging, add backport labels and re-run https://circleci.com/gh/hashicorp/consul/463501.

**Pull request merged**